### PR TITLE
udevadm: use parse_sec instead of atoi

### PR DESCRIPTION
**Pull request by stefan-it**
This PR uses the `safe_atoi` function for parsing the `--timeout` command line argument for `udevadm`.

### CODE REVIEW

File: src/udev/udevadm-control.c [hunk 1]
```diff
@@ -136,11 +137,12 @@ static int adm_control(struct udev *udev, int argc, char *argv[]) {
                 case 't': {
                         int seconds;
 
-                        seconds = atoi(optarg);
-                        if (seconds >= 0)
-                                timeout = seconds;
-                        else
+                        rc = safe_atoi(optarg, &seconds);
```
Comment (poettering) [inline src/udev/udevadm-control.c]: Hmm, I think if we rework this we should change this to use parse_sec() instead, which is a superset of atoi() in this context, and understands the usual "h", "min", ... suffixes. (of course, it actually returns usec_t, instead of int, but converting that should be simple enough...

File: src/udev/udevadm-control.c [hunk 2]
```diff
@@ -136,11 +137,12 @@ static int adm_control(struct udev *udev, int argc, char *argv[]) {
...
+
+                        if (rc < 0 || seconds < 0)
                                 fprintf(stderr, "invalid timeout value\n");
```
Comment (poettering) [inline src/udev/udevadm-control.c]: For extra bonus points change this to log_error()...

File: src/udev/udevadm-control.c (hunk 1 above)
Comment (stefan-it) [reply src/udev/udevadm-control.c]: What do you think about the following code:

``` c
case 't': {
        usec_t s;
        int seconds;
        int r;

        r = parse_sec(optarg, &s);
        if (r < 0) {
                log_error("Failed to parse timeout value '%s'.", optarg);
                return rc;
        }

        if ((s/USEC_PER_SEC) > INT_MAX)
                log_error("Timeout value is out of range.");
        else {
                seconds = (int) (s/USEC_PER_SEC);
                timeout = seconds;
                rc = 0;
        }
        break;
}
```

As the `timeout` variable is an `int` I added an overflow detection:

``` bash
[root@arch-mse systemd]# ./udevadm control --timeout="infinity"
Timeout value is out of range.
[root@arch-mse systemd]# ./udevadm control --timeout="2147483648"
Timeout value is out of range.
```

 `parse_sec` calls `parse_time` and no "negative" values are allowed, so there's no need for an underflow detection, because in this cases a "Failed to parse timeout value" message will be returned:

``` bash
[root@arch-mse systemd]# ./udevadm control --timeout="-5d"
Failed to parse timeout value '-5d'.
```

In case of no error the return value `0` would be returned (before this patch, `1` was returned).

Comment (poettering) [reply src/udev/udevadm-control.c]: the "return rc" should be "return r".

And I'd probably round up in the division... i.e.`(s+USEC_PER_SEC-1)/USEC_PER_SEC`, but I figure that doesn't matter much...

looks great otherwise...

Comment (stefan-it) [reply src/udev/udevadm-control.c]: Then we could support the `infinity` option as `--timeout` value:

``` c
seconds = s != USEC_INFINITY ? (int) ((s + USEC_PER_SEC - 1) / USEC_PER_SEC) : INT_MAX;
```

I'll replace the `fprintf` calls + push it again!

**Comment by stefan-it** [comment]
I force pushed the latest changes, but one CI build fails with:

``` bash
checking that there are no running jobs
autopkgtest [20:17:48]: test boot-smoke: -----------------------]
autopkgtest [20:17:48]: test boot-smoke:  - - - - - - - - - - results - - - - - - - - - -
boot-smoke           PASS
autopkgtest [20:17:49]: @@@@@@@@@@@@@@@@@@@@ summary
timedated            PASS
hostnamed            PASS
localed-locale       PASS
localed-x11-keymap   PASS
logind               PASS
unit-config          PASS
storage              PASS
networkd-test.py     PASS
build-login          PASS
boot-and-services    PASS
udev                 PASS
upstream             FAIL non-zero exit status 1
boot-smoke           PASS
Exit request sent.
```

**Comment by poettering** [comment]
CI failure looks unrelated. @martinpitt, any idea?

File: src/udev/udevadm-control.c [hunk 3]
```diff
@@ -134,13 +135,23 @@ static int adm_control(struct udev *udev, int argc, char *argv[]) {
...
                 case 't': {
+                        usec_t s;
                         int seconds;
+                        int r;
 
-                        seconds = atoi(optarg);
-                        if (seconds >= 0)
+                        r = parse_sec(optarg, &s);
+                        if (r < 0) {
+                                log_error("Failed to parse timeout value '%s'.", optarg);
```
Comment (poettering) [inline src/udev/udevadm-control.c]: one more improvement: shorten this to:

``` c
if (r < 0)
        return log_error_errno(r, "Failed to parse timeout value '%s'.", optarg);
```

**Comment by poettering** [comment]
looks great. one suggestion though, see above. might be worth fixing this and pushing again, maybe the CI when restarted that way cleans up itself on its own ;-)

**Comment by poettering** [comment]
lgtm

**Comment by martinpitt** [comment]
Do you still have a pointer to the previous log that has the failed i386 test?

**Comment by stefan-it** [comment]
Yes, it can be found [here](https://objectstorage.prodstack4-5.canonical.com/v1/AUTH_77e2ada1e7a84929a74ba3b87153c0ac/autopkgtest-xenial-pitti-systemd-semaphore/xenial/amd64/s/systemd-upstream/20161010_201800@/log.gz) :)

**Comment by martinpitt** [comment]
All tests green now, landing. Thanks!

So in that failed run, `test/TEST-03-JOBS/test.sh` failed, but not with a visible error message -- all units apparently started ("OK") during boot. The journal looks truncated though, it just contains some block with tons of `transport endpoint not connected` which is irrelevant, I believe. So I'm afraid there's not much more we can squeeze out of that. Also, first time this fails like this, so probably not worth spending much time on (if it becomes reproducible, we can have a closer look).

**Comment by evverx** [comment]
@martinpitt 

> Also, first time this fails like this, so probably not worth spending much time on

I have already seen: https://github.com/systemd/systemd/pull/4319#issuecomment-252523357
But I have no idea what is going on
